### PR TITLE
Client has a main game loop

### PR DESCRIPTION
**Pull request by mprtcz**
In this pull req there are some tests for fx Controller class and a main game loop.

### CODE REVIEW

File: src/main/java/com/spanish_inquisition/battleship/client/MainMenuController.java [hunk 1]
```diff
@@ -26,46 +26,48 @@
 public class MainMenuController {
 
     @FXML
-    public VBox centralVBox;
+    VBox centralVBox;
```
Comment (piotr921) [inline src/main/java/com/spanish_inquisition/battleship/client/MainMenuController.java]: 👍

File: src/main/java/com/spanish_inquisition/battleship/client/MainMenuController.java [hunk 2]
```diff
@@ -26,46 +26,48 @@
...
     @FXML
-    public TextField nameTextField;
+    TextField nameTextField;
```
Comment (piotr921) [inline src/main/java/com/spanish_inquisition/battleship/client/MainMenuController.java]: 🥅

File: src/main/java/com/spanish_inquisition/battleship/client/MainMenuController.java [hunk 3]
```diff
@@ -90,22 +92,20 @@ public void onNameTextFieldEntered() {
...
+    private void buildPlayerBoard() {
         Platform.runLater(() -> playersLabel.setText("Set Up your ships"));
-        this.game = new Game();
         this.game.buildPlayersBoard(new BoardController(new GameBoard(this.playersGridPane)));
         this.fleetSetupButton.setVisible(true);
     }
 
     @FXML
-    public void onFleetSetupButtonClicked(){
+    public void onFleetSetupButtonClicked() {
```
Comment (piotr921) [inline src/main/java/com/spanish_inquisition/battleship/client/MainMenuController.java]: 👍

File: src/main/java/com/spanish_inquisition/battleship/client/game/Game.java [hunk 4]
```diff
@@ -10,19 +14,60 @@
...
+        this.responsesBus = socketClient.getResponsesBus();
+    }
+
     public void placePlayersShips() {
         logger.log(DEFAULT_LEVEL, "Setting fleet for player");
         boardController.placeShips();
     }
 
-    public String getShipPlacementForServer(){
+    public String getShipPlacementForServer() {
```
Comment (piotr921) [inline src/main/java/com/spanish_inquisition/battleship/client/game/Game.java]: 💩 i like it

File: src/main/java/com/spanish_inquisition/battleship/client/game/Game.java [hunk 5]
```diff
@@ -10,19 +14,60 @@
...
         return boardController.getMessageForServer();
     }
+
+    public void runTheGame() throws InterruptedException {
+        if (this.socketClient != null) {
+            while (true) {
+                    Thread.sleep(GAME_LOOP_SLEEP);
+                if (this.responsesBus.hasServerResponses()) {
+                    NetworkMessage message = this.responsesBus.getAServerResponse();
```
Comment (piotr921) [inline src/main/java/com/spanish_inquisition/battleship/client/game/Game.java]: maybe getServerResponse

File: src/main/java/com/spanish_inquisition/battleship/client/network/ResponsesBus.java [hunk 6]
```diff
@@ -18,11 +18,11 @@
         this.serverResponses = new ConcurrentLinkedQueue<>();
     }
 
-    boolean hasServerResponses() {
+    public boolean hasServerResponses() {
         return !this.serverResponses.isEmpty();
     }
 
-    NetworkMessage getAServerResponse() {
+    public NetworkMessage getAServerResponse() {
```
Comment (piotr921) [inline src/main/java/com/spanish_inquisition/battleship/client/network/ResponsesBus.java]: I would remove 'A' from method name.

File: src/test/java/com/spanish_inquisition/battleship/client/MainMenuControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    @Test
+    public void testOnFeatureButtonClicked() throws Exception {
+        MainMenuController mainMenuController = getControllerInstance();
+        mainMenuController.nameTextField.setText("DummyText");
+        assertTrue(mainMenuController.playerNameVBox.isVisible());
+        int threadCount = Thread.activeCount();
+        mainMenuController.onFeatureButtonClicked();
+        assertEquals(threadCount + 1, Thread.activeCount());
+        assertFalse(mainMenuController.playerNameVBox.isVisible());
+    }
```
Comment (piotr921) [inline src/test/java/com/spanish_inquisition/battleship/client/MainMenuControllerTest.java]: Given,  When, Then.

File: src/main/java/com/spanish_inquisition/battleship/client/game/Game.java [hunk 8]
```diff
@@ -10,19 +14,60 @@
...
+                    System.out.println("s = " + message);
+                    if (message == null) {
+                        continue;
+                    }
+                    if (message.getHeader() == Header.PLAYER_TURN) {
+                        //notify about his move
+                    } else if (message.getHeader() == Header.OPPONENT_TURN) {
+                        // notify that it is not his turn
```
Comment (Oplia512) [inline src/main/java/com/spanish_inquisition/battleship/client/game/Game.java]: it is better to wrap it in switch-case

File: src/main/java/com/spanish_inquisition/battleship/client/game/Game.java [hunk 9]
```diff
@@ -10,19 +14,60 @@
...
     }
 
-    public String getShipPlacementForServer(){
+    public String getShipPlacementForServer() {
         return boardController.getMessageForServer();
     }
+
+    public void runTheGame() throws InterruptedException {
+        if (this.socketClient != null) {
+            while (true) {
```
Comment (arturAR) [inline src/main/java/com/spanish_inquisition/battleship/client/game/Game.java]: risky loop

File: src/main/java/com/spanish_inquisition/battleship/client/game/Game.java [hunk 10]
```diff
@@ -10,19 +14,60 @@
...
+                    if (message.getHeader() == Header.PLAYER_TURN) {
+                        //notify about his move
+                    } else if (message.getHeader() == Header.OPPONENT_TURN) {
+                        // notify that it is not his turn
+                    } else if (Header.isResponseFieldChanging(message.getHeader())) {
+                        // make changes to the opponent's board
+                    } else if (message.getHeader() == Header.GAME_WON) {
+                        break;
+                        // notify the player he won or lost
+                    }
```
Comment (Oplia512) [inline src/main/java/com/spanish_inquisition/battleship/client/game/Game.java]: a lot of if here have to be refactored

File: src/main/java/com/spanish_inquisition/battleship/client/game/Game.java [hunk 11]
```diff
@@ -10,19 +14,60 @@
...
+                }
+            }
+        }
+    }
+
+    public void acceptPlayersName(String name) {
+        this.name = name;
+        if (socketClient != null) {
```
Comment (Oplia512) [inline src/main/java/com/spanish_inquisition/battleship/client/game/Game.java]: one million $$$

File: src/test/java/com/spanish_inquisition/battleship/client/MainMenuControllerTest.java [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+        final CountDownLatch latch = new CountDownLatch(1);
+        SwingUtilities.invokeLater(() -> {
+            new JFXPanel(); // initializes JavaFX environment
+            latch.countDown();
+        });
+    }
+
+    @Test
+    public void testOnFeatureButtonClicked() throws Exception {
+        MainMenuController mainMenuController = getControllerInstance();
```
Comment (arturAR) [inline src/test/java/com/spanish_inquisition/battleship/client/MainMenuControllerTest.java]: Empty line to divide part?

File: src/test/java/com/spanish_inquisition/battleship/client/MainMenuControllerTest.java [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+        mainMenuController.onFeatureButtonClicked();
+        assertEquals(threadCount + 1, Thread.activeCount());
+        assertFalse(mainMenuController.playerNameVBox.isVisible());
+    }
+
+    @Test
+    public void testOnNameTextFieldEntered() throws Exception {
+        MainMenuController mainMenuController = getControllerInstance();
+        mainMenuController.nameTextField.setText("DummyText");
+        assertTrue(mainMenuController.playerNameVBox.isVisible());
```
Comment (arturAR) [inline src/test/java/com/spanish_inquisition/battleship/client/MainMenuControllerTest.java]: Empty line to divide part?

File: src/test/java/com/spanish_inquisition/battleship/client/MainMenuControllerTest.java [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+        mainMenuController.socketClient = socketClient;
+        assertFalse(mainMenuController.fleetSetupButton.isDisabled());
+        when(mainMenuController.game.getShipPlacementForServer()).thenReturn("");
+        mainMenuController.onSendToServerButtonClicked();
+        assertTrue(mainMenuController.fleetSetupButton.isDisabled());
+    }
+
+    private MainMenuController getControllerInstance() {
+        Game game = mock(Game.class);
+        MainMenuController mainMenuController = new MainMenuController();
```
Comment (arturAR) [inline src/test/java/com/spanish_inquisition/battleship/client/MainMenuControllerTest.java]: Empty line to divide part?

File: src/test/java/com/spanish_inquisition/battleship/client/MainMenuControllerTest.java [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+        mainMenuController.sendToServerButton.setVisible(false);
+        mainMenuController.onFleetSetupButtonClicked();
+        assertTrue(mainMenuController.sendToServerButton.isVisible());
+    }
+
+    @Test
+    public void testOnSendToServerButtonClicked() throws Exception {
+        SocketClient socketClient = mock(SocketClient.class);
+        MainMenuController mainMenuController = getControllerInstance();
+        mainMenuController.socketClient = socketClient;
```
Comment (arturAR) [inline src/test/java/com/spanish_inquisition/battleship/client/MainMenuControllerTest.java]: Empty line to divide part?

**Pull request merged**